### ISSUE TITLE
Transformer masking

### ISSUE THREAD

**Issue by derekdriggs** [facebookresearch/CovidPrognosis#16]
I have a question about how the transformer layer treats sequences of images of different lengths. Consider the following example evaluation of a MIPModel:

image_model = DenseNet()
net = MIPModel(image_model, 1024, 64, 2, 2, 1024, 0.1, 0.1, 'sum')
B1, N1, C1, H1, W1 = 2, 4, 3, 224, 224
images = torch.rand(B1, N1, C1, H1, W1)
times = torch.tensor([
    [-30, -20, -10, 0],
    [-30, -20, 0, 0],      # Extra 0 for padding
])
lens = torch.tensor([4, 3])
x = net(images, times, lens)

In this batch of size 2, the first sequence has 4 images, and the second sequence has 3 images. We add padding to the times corresponding to the shorter sequence, and we pad the shorter sequence of images with a "random" image so they are both of length 4.

This "padding" image should not be considered by the transformer. However, the function "_apply_transformer" in MIPModel does not mask this image for the transformer. In fact, "lens" is input into "_apply_transformer", but it is not used. "Lens" is used to drop the extra images in "image_feats_combined" during the "_pool" function, but the transformer has already used the "padding" images to create "image_feats_trans".

Is this correct, and if so, how should we pad image sequences of different lengths so that the "padding" images do not affect the transformer output?

**Comment by derekdriggs** [comment]
Apologies, but are there any updates on how we should pad sequences and mask these images for the transformer? I very much appreciate your help with this.

**Comment by mmuckley** [comment]
Hello @derekdriggs, sorry about the long reply. I would try to help but this is a fairly specific question and I don't want to provide incorrect information. @anuroopsriram has been very busy preparing for NeurIPS lately. I'll try to follow up with him again after the deadline.

**Comment by derekdriggs** [comment]
Of course, please take your time. Thank you again for helping with this issue and for sharing your research.

Instead of a complete solution to the problem, I suppose I would just like to know if there is a mistake in my understanding, or if there is a mistake in "_apply_transformer".

It seems to me that "lens" is meant to be used to mask the padding images, but it is not used at all, and these images are never masked, affecting the transformer output.

**Comment by anuroopsriram** [comment]
Hi @derekdriggs: Thanks for pointing this out. The sequences do need to be masked. I think the masking code got lost in the final version of the code. I don't have time right now, but I can add masking in a few weeks. 
In the meantime, if you are interested in adding this, I would be happy to review it.

**Comment by derekdriggs** [comment]
Hi anuroopsriram, thank you very much for your advice, I was noticing the SIP classifier was outperforming the MIP classifier on our data, and masking seemed to help. I will submit our minor adjustments soon and hope it is useful for others as well.
